File: frontend/src/types/arena.ts
export type PriceLevel = {
  agent_id?: string;
  owner?: "normal" | "abuser" | string;
  price: number;
  quantity: number;
  scenario_id?: string;
  scenario_name?: string;
};

export type OrderBookSnapshot = {
  bids: PriceLevel[];
  asks: PriceLevel[];
  best_bid: number | null;
  best_ask: number | null;
  mid: number | null;
  spread: number | null;
};

export type MarketFeatures = {
  spread_bps: number;
  depth_top_n: number;
  imbalance: number;
  order_book_imbalance?: number;
  top_n_bid_depth?: number;
  top_n_ask_depth?: number;
  message_rate: number;
  message_rate_per_sec?: number;
  cancel_to_trade_ratio: number;
  order_lifetime_ms: number;
  wall_size_ratio: number;
  depth_change_pct: number;
};

export type DetectorScore = {
  name: string;
  confidence: number;
  alert: boolean;
  severity?: "low" | "medium" | "high" | "critical";
  evidence?: EvidenceItem[];
};

export type DetectorScores = {
  scores: DetectorScore[];
  alerts: DetectorScore[];
};

export type AgentEvent = {
  type: string;
  timestamp?: number;
  order_id?: string;
  aggressor_order_id?: string;
  resting_order_id?: string;
  agent_id?: string;
  aggressor_agent_id?: string;
  resting_agent_id?: string;
  side?: "buy" | "sell";
  price?: number;
  quantity?: number;
  scenario_id?: string | null;
  scenario_name?: string | null;
  scenario_family?: string | null;
  [key: string]: unknown;
};

export type EvidenceItem = {
  key: string;
  label: string;
  value: string | number | boolean;
  unit?: string;
  interpretation?: string;
};

export type Incident = {
  id: string;
  title: string;
  type: string;
  agent: string;
  confidence: number;
  severity: "Low" | "Medium" | "High" | "Critical";
  evidence: EvidenceItem[];
  explanation: string;
  scenario_id?: string;
  scenario_family?: string;
};

export type AttackStage =
  | "armed"
  | "wall_placed"
  | "pressure_phase"
  | "wall_cancelled"
  | "cancelled"
  | "incident_confirmed"
  | "done";

export type AttackStageStatus = "pending" | "active" | "completed";

export type AttackStageSnapshot = {
  detector_confidence?: number;
  label: string;
  stage: AttackStage;
  status: AttackStageStatus;
  tick?: number;
  timestamp?: number;
};

export type AttackTrackerState = {
  scenario_id: string;
  scenario_name: string;
  scenario_family: string;
  agent_id: string;
  current_stage?: AttackStage;
  start_tick: number;
  stages?: AttackStageSnapshot[];
  status: AttackStage | string;
  evidence?: EvidenceItem[];
};

export type ArenaState = {
  tick: number;
  running: boolean;
  events: AgentEvent[];
  book: OrderBookSnapshot;
  best_bid: number | null;
  best_ask: number | null;
  mid: number | null;
  spread: number | null;
  active_agents: string[];
  active_scenario: AttackTrackerState | null;
  detectors: DetectorScores;
  incidents?: Incident[];
  features?: Partial<MarketFeatures>;
};

export type ArenaWebSocketMessage = {
  type: "arena_state";
  version: number;
  timestamp: string;
  payload: ArenaState;
};

export type ScenarioConfig = {
  label: string;
  slug: string;
  scenario_family: string;
  agent_id: string;
  description: string;
  launch_endpoint?: string;
  parameters?: Record<string, unknown>;
  market_regime?: "calm" | "volatile" | "thin_liquidity";
  goal?: "obvious" | "stealth" | "hard_to_detect";
  source?: "nebius" | "mock";
  expected_detector_risk?: number;
  safety_note?: string;
};

export type BenchmarkResult = {
  avg_detection_latency_ms?: number;
  scenario: string;
  precision: number;
  recall: number;
  f1: number;
};

File: backend/app/schemas/arena.py
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class ScenarioType(StrEnum):
    SPOOFING_LIKE_WALL = "spoofing_like_wall"
    LAYERING_LIKE = "layering_like"
    QUOTE_STUFFING = "quote_stuffing"
    LIQUIDITY_EVAPORATION = "liquidity_evaporation"
    PANIC_SELLOFF = "panic_selloff"


class SimulationStatus(StrEnum):
    STOPPED = "stopped"
    RUNNING = "running"
    PAUSED = "paused"


class MarketRegime(StrEnum):
    CALM = "calm"
    VOLATILE = "volatile"
    THIN_LIQUIDITY = "thin_liquidity"


class RedTeamGoal(StrEnum):
    OBVIOUS = "obvious"
    STEALTH = "stealth"
    HARD_TO_DETECT = "hard_to_detect"


class RedTeamScenarioGenerateRequest(BaseModel):
    scenario_family: str
    market_regime: MarketRegime
    goal: RedTeamGoal
    constraints: dict[str, Any] = Field(default_factory=dict)


class ScenarioConfig(BaseModel):
    label: str
    slug: ScenarioType
    scenario_family: str
    agent_id: str
    description: str
    launch_endpoint: str
    parameters: dict[str, Any] = Field(default_factory=dict)
    market_regime: MarketRegime
    goal: RedTeamGoal
    source: Literal["nebius", "mock"]
    expected_detector_risk: float = Field(ge=0.0, le=1.0)
    safety_note: str


class BenchmarkResult(BaseModel):
    scenario: str
    precision: float
    recall: float
    f1: float
    avg_detection_latency_ms: float | None = None


class PriceLevel(BaseModel):
    price: float
    quantity: float
    agent_id: str | None = None
    owner: str | None = None
    scenario_id: str | None = None
    scenario_name: str | None = None


class OrderBookSnapshot(BaseModel):
    bids: list[PriceLevel]
    asks: list[PriceLevel]
    best_bid: float | None
    best_ask: float | None
    mid: float | None
    spread: float | None


class MarketFeatures(BaseModel):
    spread_bps: float
    depth_top_n: float
    imbalance: float
    message_rate: float
    cancel_to_trade_ratio: float
    order_lifetime_ms: float
    wall_size_ratio: float
    depth_change_pct: float
    distance_from_touch_bps: float = 0.0
    cancel_probability: float = 0.0
    execution_ratio: float = 0.0
    replenishment_rate: float = 0.0
    side_switching_rate: float = 0.0
    participant_order_linkage: float = 0.0
    linked_participant_ids: str = ""
    linked_order_ids: str = ""
    linked_event_ids: str = ""
    large_level_count: float = 0.0


class DetectorScore(BaseModel):
    name: str
    confidence: float = Field(ge=0.0, le=1.0)
    alert: bool
    severity: Literal["low", "medium", "high", "critical"] | None = None
    evidence: list["EvidenceItem"] | None = None


class DetectorScores(BaseModel):
    scores: list[DetectorScore]
    alerts: list[DetectorScore]


class AgentEvent(BaseModel):
    model_config = ConfigDict(extra="allow")

    type: str
    event_id: str | None = None
    timestamp: float | None = None
    order_id: str | None = None
    aggressor_order_id: str | None = None
    resting_order_id: str | None = None
    agent_id: str | None = None
    aggressor_agent_id: str | None = None
    resting_agent_id: str | None = None
    side: Literal["buy", "sell"] | None = None
    price: float | None = None
    quantity: float | None = None
    scenario_id: str | None = None
    scenario_name: str | None = None
    scenario_family: str | None = None


class EvidenceItem(BaseModel):
    key: str
    label: str
    value: str | int | float | bool
    unit: str | None = None
    interpretation: str | None = None


class Incident(BaseModel):
    id: str
    title: str
    type: str
    agent: str
    confidence: float = Field(ge=0.0, le=1.0)
    severity: Literal["Low", "Medium", "High", "Critical"]
    evidence: list[EvidenceItem]
    explanation: str
    scenario_id: str | None = None
    scenario_family: str | None = None


class AttackStage(StrEnum):
    ARMED = "armed"
    WALL_PLACED = "wall_placed"
    PRESSURE_PHASE = "pressure_phase"
    WALL_CANCELLED = "wall_cancelled"
    CANCELLED = "cancelled"
    INCIDENT_CONFIRMED = "incident_confirmed"
    DONE = "done"


class AttackStageSnapshot(BaseModel):
    detector_confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    label: str
    stage: AttackStage
    status: Literal["pending", "active", "completed"]
    tick: int | None = None
    timestamp: float | None = None


class ScenarioLabel(BaseModel):
    label_id: str
    run_id: str
    scenario_id: str
    scenario_family: str
    scenario_name: str
    seed: int
    start_tick: int
    expected_end_tick: int | None = None
    actual_end_tick: int | None = None
    agent_ids: list[str]
    event_ids: list[str] = Field(default_factory=list)
    order_ids: list[str] = Field(default_factory=list)
    manipulation_windows: list[dict[str, int | None]] = Field(default_factory=list)
    phase_windows: dict[str, dict[str, int | None]] = Field(default_factory=dict)
    parameters: dict[str, Any] = Field(default_factory=dict)


class AttackTrackerState(BaseModel):
    scenario_id: str
    scenario_name: str
    scenario_family: str
    agent_id: str
    current_stage: AttackStage | None = None
    start_tick: int
    stages: list[AttackStageSnapshot] | None = None
    status: AttackStage | str
    evidence: list[EvidenceItem] | None = None
    label: ScenarioLabel | None = None


class ArenaState(BaseModel):
    tick: int
    running: bool
    events: list[AgentEvent]
    book: OrderBookSnapshot
    best_bid: float | None
    best_ask: float | None
    mid: float | None
    spread: float | None
    active_agents: list[str]
    active_scenario: AttackTrackerState | None
    detectors: DetectorScores
    incidents: list[Incident] | None = None
    features: dict[str, Any] | MarketFeatures | None = None
